crates/ibc/src/applications/transfer/coin.rs

```diff
@@ -59,24 +71,30 @@ where
         // a letter, a number or a separator ('/', ':', '.', '_' or '-').
         // Loosely copy the regex from here:
         // https://github.com/cosmos/cosmos-sdk/blob/v0.45.5/types/coin.go#L760-L762
-        let matcher = regex!(br"([0-9]+)([a-zA-Z0-9/:\\._\x2d]+)");
-
-        let (m1, m2) = matcher.match_slices(coin_str.as_bytes()).ok_or_else(|| {
-            TokenTransferError::InvalidCoin {
+        //
+        // equivalent regex code in rust:
+        // let re = Regex::new(r"^(?<amount>[0-9]+)(?<denom>[a-zA-Z0-9/:._-]+)$")?;
+        // let cap = re.captures("123stake")?;
+        // let (amount, denom) = (cap.name("amount")?.as_str(), cap.name("denom")?.as_str());
+
+        let (amount, denom) = coin_str
+            .chars()
+            .position(|x| !x.is_numeric())
+            .map(|index| coin_str.split_at(index))
+            .filter(|(amount, denom)| !amount.is_empty() && !denom.is_empty())
+            .filter(|(_, denom)| {
+                denom.chars().all(|x| {
+                    matches!(x, 'a'..='z' | 'A'..='Z' | '0'..='9' | '/' | ':' | '.' | '_' | '-')
+                })
+            })
```

Comment:
Here we can simplify the implementation and use the same approach as [here](https://github.com/cosmos/ibc-rs/blob/dc8e4161dd5e00a2031cf3feb58dad1ed7e26bdb/crates/ibc/src/core/ics24_host/identifier/validate.rs#L22-L31) by defnining e.g. `VALID_DENOM_CHARACTERS` , after which the `!denom.is_empty()` check won’t be needed anymore too. Even we might be able to decouple checks and filter of the amount from the denom. Appreciate a further look on this.

Will looks like something like this:

```rust
const VALID_DENOM_CHARACTERS: &str = "/:._-";

let (amount, denom) = coin_str
            .chars()
            .position(|x| !x.is_numeric())
            .map(|index| coin_str.split_at(index))
            .filter(|(amount, _)| !amount.is_empty())
            .filter(|(_, denom)| {
                denom
                    .chars()
                    .all(|x| x.is_alphanumeric() || VALID_DENOM_CHARACTERS.contains(x))
            })
            .ok_or_else(|| TokenTransferError::InvalidCoin {
                coin: coin_str.to_string(),
            })?;
```